src/modules/Game.class.js

```diff
@@ -1,68 +1,270 @@
+
+      while (row.length < this.state[i].length) {
+        row.push(0);
+      } // [4, 4, 0, 0]
+
+      this.state[i] = row;
+    }
+
+    this.addRandomTile();
+
+    if (!this.movesAvailable()) {
+      this.status = Game.STATUS_LOSE;
+    }
+  }
+
+  moveRight() {
+    if (this.status === Game.STATUS_WIN || this.status === Game.STATUS_LOSE) {
+      return;
+    }
+
+    // [0, 4, 2, 2]
+    for (let i = 0; i < this.state.length; i++) {
+      let row = this.state[i].filter((val) => val !== 0); // [4, 2, 2]
+
+      for (let j = row.length - 1; j > 0; j--) {
+        if (row[j] === row[j - 1]) {
+          row[j] *= 2;
+          row[j - 1] = 0;
+
+          if (row[j] === Game.WIN_NUMBER) {
+            this.status = Game.STATUS_WIN;
+          }
+
+          this.score += row[j]; // [4, 0, 4]
+        }
+      }
+
+      row = row.filter((val) => val !== 0); // [4, 4]
+
+      while (row.length < this.state[i].length) {
+        row.unshift(0);
+      } // [0, 0, 4, 4]
+
+      this.state[i] = row;
+    }
+
+    this.addRandomTile();
+
+    if (!this.movesAvailable()) {
+      this.status = Game.STATUS_LOSE;
+    }
   }
 
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  moveUp() {
+    if (this.status === Game.STATUS_WIN || this.status === Game.STATUS_LOSE) {
+      return;
+    }
+
+    for (let col = 0; col < this.state[0].length; col++) {
+      let column = [];
+
+      for (let row = 0; row < this.state.length; row++) {
+        if (this.state[row][col] !== 0) {
+          column.push(this.state[row][col]);
+        }
+      }
+
+      for (let i = 0; i < column.length - 1; i++) {
+        if (column[i] === column[i + 1]) {
+          column[i] *= 2;
+          column[i + 1] = 0;
+
+          if (column[i] === Game.WIN_NUMBER) {
+            this.status = Game.STATUS_WIN;
+          }
+
+          this.score += column[i];
+        }
+      }
+
+      column = column.filter((val) => val !== 0);
+
+      while (column.length < this.state.length) {
+        column.push(0);
+      }
+
+      for (let row = 0; row < this.state.length; row++) {
+        this.state[row][col] = column[row];
+      }
+    }
+
+    this.addRandomTile();
+
+    if (!this.movesAvailable()) {
+      this.status = Game.STATUS_LOSE;
+    }
+  }
+
+  moveDown() {
+    if (this.status === Game.STATUS_WIN || this.status === Game.STATUS_LOSE) {
```

Comment:
Maybe you shouldn't do anything when the status is IDLE as well. See the error in the comments